dist_ir/backend/torch.py

```diff
@@ -397,34 +451,77 @@ def run_function(
         value_map[v] = x
     assert len(fn.inputs) == len(inputs)
 
-    def print_memory_usage():
-        t = torch.cuda.get_device_properties(0).total_memory
-        r = torch.cuda.memory_reserved(0)
-        a = torch.cuda.memory_allocated(0)
-        print(f"Total: {t} Reserved: {r} Allocated: {a} Free: {r-a}")
+    if ctx.measure_peak_memory:
+        torch.cuda.synchronize(device=rank)
+        memory_usage = get_memory_usage(rank)
+        allocated_memory = memory_usage.allocated
+        peak_memory = allocated_memory
+    else:
+        peak_memory = 0.0
+
+    recv_output_values = set()
+    op_runtime_events = []
 
     # Run ops
     for op in fn.ops:
         inputs = tuple(value_map[v] for v in op.inputs)
         kwargs = {} if op.attributes is None else {**op.attributes}
         kwargs["ctx"] = ctx
 
+        if record_op_runtimes:
+            add_event(ctx, op_runtime_events)
+        if op.op_type == "RecvP2P":
```

Comment:
I think the problem is we wouldn't know inside `_recv` which buffer belongs to that op since currently the op itself is the key in `recv_buffers` and we aren't passing any op ID to the individual functions.

Reply:
I see. So to do this cleanly, we'd either have to pass the op to every implementation function like `_recv`, or give each `RecvP2P` op a new attribute while projection which contains some unique identifier? Perhaps we can leave a TODO to clean this up at some point.

Reply:
That's right yea, I'll leave a TODO